File: javaprojects/common/src/main/java/minigames/commands/CommandPackage.java
package minigames.commands;

import java.util.List;
import io.vertx.core.json.JsonObject;

/**
 * A package of commands sent from a player to a game.
 * 
 * Commands are just given as a list of JSON objects. 
 * This lets GameClients and GameServers implement bespoke commands at will.
 */
public record CommandPackage(
    String gameServer,
    String gameId,
    String player,
    List<JsonObject> commands
) {
    
    public static CommandPackage fromJson(JsonObject json) {
        return new CommandPackage(
            json.getString("gameServer"),
            json.getString("gameId"),
            json.getString("player"),
            json.getJsonArray("commands").stream().map((o) -> (JsonObject)o).toList()
        );
      }
}

File: scalaprojects/scalafx-client/src/main/scala/minigames/commands/CommandPackage.scala
package minigames.commands

import io.vertx.core.json.JsonObject
import scala.jdk.CollectionConverters.*
import io.vertx.core.json.JsonArray

case class CommandPackage(
    gameServer:String,
    gameId:String,
    player:String,
    commands:Seq[JsonObject]
) {

    def toJson = JsonObject()
        .put("gameServer", gameServer)
        .put("gameId", gameId)
        .put("player", player)
        .put("commands", {
            commands.foldLeft(JsonArray()) { _.add(_) }
        })

}

object CommandPackage {
    def fromJson(json:JsonObject) = CommandPackage(
        gameServer=json.getString("gameServer"),
        gameId=json.getString("gameId"),
        player=json.getString("player"),
        commands=json.getJsonArray("commands").toSeq[JsonObject]
    )
}